### PR TITLE
fix: respect package.json when installing bare specifiers, fix some type bugs

### PR DESCRIPTION
**Pull request by Bubblyworld**
When resolving a package in `generator.install(...)`, if there are no version
constraints then the constraints in the `package.json` are used, rather than
the current behaviour which is to use the latest version.

Also fixes a bug where `InstallTarget`s were being put in the installer's
constraint map instead of `PackageTarget`s, which caused a crash during my local
testing.

### CODE REVIEW

**Comment by Bubblyworld** [comment]
By the way, there's a rare race condition in `tests/html/multi.test.js` that seems to have been around for at least since `41bcad4`. Causes intermittent failures, haven't managed to isolate the root cause yet.

File: src/install/lock.ts [hunk 1]
```diff
@@ -41,15 +35,15 @@ export interface LockResolutions {
   };
 }
 
+interface PackageTargetMap {
+  [pkgName: string]: PackageTarget | URL,
```
Comment (Bubblyworld) [inline src/install/lock.ts:39]: This is the interface actually used by the code - I've fixed all the type issues related to `VersionConstraints`, it was causing some weird bugs where the type doesn't match the actual input.

File: src/install/package.ts [hunk 2]
```diff
@@ -207,6 +207,30 @@ export async function parseTarget(
   if (versionIndex !== -1) alias = pkg.pkgName.slice(0, versionIndex);
   else alias = pkg.pkgName;
 
+  // If no version is specified, we fallback to the constraints in the parent
+  // package config if they exist:
+  const pcfg = await resolver.getPackageConfig(parentPkgUrl.href);
```
Comment (Bubblyworld) [inline src/install/package.ts:212]: Main change, if there's no version specified in a `generator.install(<spec>, ...)` call, then this looks up version constraints from the base `package.json` (if it exists). I think this is actually the root cause of a few bugs, including #211.

**Comment by guybedford** [comment]
Here's a case that isn't working for this:

package.json
```
{
  "name": "app",
  "exports": "./app.js",
  "dependencies": {
    "es-module-lexer": "^1"
  }
}
```

importmap.json
```
{
  "imports": {
    "es-module-lexer": "https://ga.jspm.io/npm:es-module-lexer@0.10.5/dist/lexer.js"
  }
}
```

Then running `node ../jspm.js link app -o sub/test.html` gives the HTML:

```
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>JSPM example</title>
    <script async src="https://ga.jspm.io/npm:es-module-shims@1.6.3/dist/es-module-shims.js" crossorigin="anonymous"></script>
<script type="importmap">
{
  "imports": {
    "app": "file:///C:/Users/Guy/Projects/jspm/a/app.js"
  },
  "scopes": {
    "file:///": {
      "es-module-lexer": "https://ga.jspm.io/npm:es-module-lexer@1.1.1/dist/lexer.js"
    }
  }
}
</script>
  </head>
  <body>
  </body>
</html>
```

This surfaces a couple of issues:

* Not rebasing the import map during the `extractMap` / `getMap` calls to the output map mapURL.
* Not reusing the older dependency from the lock file, when link should ideally be fully locking

**Pull request merged**